### PR TITLE
Run code formatter on lib/elixir/lib/kernel.ex

### PR DESCRIPTION
**Pull request by fabrik42**
Ok, I got the Kernel 😰 If this is totally wrong, please let me know.

One thing that I recognized is that the Kernel contains a lot of very long strings with interpolations, used for warnings, error messages, etc.

It seems that three different kind of ways are used to construct this strings:

* [String concatenation](https://github.com/elixir-lang/elixir/compare/master...fabrik42:exfmt-kernel?expand=1#diff-e06d5ebc2907d951188e88476f85af51R4174)
* [Using binary type](https://github.com/elixir-lang/elixir/compare/master...fabrik42:exfmt-kernel?expand=1#diff-e06d5ebc2907d951188e88476f85af51R3413)
* [New lines in interpolation](https://github.com/elixir-lang/elixir/compare/master...fabrik42:exfmt-kernel?expand=1#diff-e06d5ebc2907d951188e88476f85af51R1714)

Would it make sense to try to streamline this as much as possible?

### CODE REVIEW

**Comment by josevalim** [comment]
I will review it soon but please note we ask people to submit only one
formatted PR per day so we have time to review it all. :)
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by fabrik42** [comment]
Ah, sorry, didn't see this. Feel free to review/close the PR whenever you are ready.

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -2593,25 +2636,42 @@ defmodule Kernel do
...
+        :elixir_errors.warn(env.line, env.file, warn_message)
 
       # TODO: Remove :compile check once on 2.0 as we no longer
       # need to warn on parse transforms in Module.put_attribute.
       name == :compile ->
         {stack, _} = :elixir_quote.escape(env_stacktrace(env), false)
-        quote do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg),
-                                       unquote(stack), unquote(line))
+
+        quote(
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please convert it to a do/end block.

File: lib/elixir/lib/kernel.ex [hunk 2]
```diff
@@ -2593,25 +2636,42 @@ defmodule Kernel do
...
+            unquote(arg),
+            unquote(stack),
+            unquote(line)
+          )
+        )
 
       :lists.member(name, [:moduledoc, :typedoc, :doc]) ->
         {stack, _} = :elixir_quote.escape(env_stacktrace(env), false)
         arg = {env.line, arg}
-        quote do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg),
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Same here.

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -2593,25 +2636,42 @@ defmodule Kernel do
...
+            unquote(line)
+          )
+        )
 
       true ->
-        quote do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg),
-                                       nil, unquote(line))
+        quote(
+          do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg), nil, unquote(line))
+        )
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Same here.

File: lib/elixir/lib/kernel.ex [hunk 4]
```diff
@@ -2746,17 +2810,21 @@ defmodule Kernel do
...
-    end)
+    )
   end
 
   defp build_if(_condition, _arguments) do
-    raise(ArgumentError, "invalid or duplicate keys for if, only \"do\" " <>
-      "and an optional \"else\" are permitted")
+    raise(
+      ArgumentError,
+      "invalid or duplicate keys for if, only \"do\" " <> "and an optional \"else\" are permitted"
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please assign the message to a variable in the previous line and remove concatenation.

File: lib/elixir/lib/kernel.ex [hunk 5]
```diff
@@ -2799,8 +2867,11 @@ defmodule Kernel do
   end
 
   defp build_unless(_condition, _arguments) do
-    raise(ArgumentError, "invalid or duplicate keys for unless, only \"do\" " <>
-      "and an optional \"else\" are permitted")
+    raise(
+      ArgumentError,
+      "invalid or duplicate keys for unless, only \"do\" " <>
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please assign the message to a variable in the previous line and also remove the parens around raise.

File: lib/elixir/lib/kernel.ex (hunk 4 above)
Comment (josevalim) [reply lib/elixir/lib/kernel.ex]: Also remove the parens around raise.

File: lib/elixir/lib/kernel.ex [hunk 6]
```diff
@@ -3039,18 +3109,25 @@ defmodule Kernel do
...
-                              <<"piping into a unary operator is deprecated, please use the ",
-                                "qualified name. For example, Kernel.+(5), instead of +5">>)
-        _ ->
-          :ok
-      end
 
-      Macro.pipe(acc, x, pos)
-    end, h, t
+    :lists.foldl(
+      fn {x, pos}, acc ->
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please extract the function out to its own variable.

File: lib/elixir/lib/kernel.ex [hunk 7]
```diff
@@ -3174,36 +3252,44 @@ defmodule Kernel do
         case ensure_evaled(list, {0, []}, expand) do
           {[head | tail], {_, []}} ->
             in_var(in_module?, left, &in_list(&1, head, tail, expand, list, in_module?))
+
           {[head | tail], {_, vars_values}} ->
             {vars, values} = :lists.unzip(:lists.reverse(vars_values))
+
             quote do
               {unquote_splicing(vars)} = {unquote_splicing(values)}
-              unquote(in_var(in_module?, left, &in_list(&1, head, tail, expand, list, in_module?)))
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please extra `&in_list(&1, head, tail, expand, list, in_module?)` to a variable outside of the `quote` (two lines above) and then use it here.

File: lib/elixir/lib/kernel.ex [hunk 8]
```diff
@@ -3213,26 +3299,33 @@ defmodule Kernel do
...
-    :lists.mapfoldl(fn
-      {:|, meta, [head, tail]}, acc ->
-        {head, acc} = ensure_evaled_element(head, acc)
-        {tail, acc} = ensure_evaled_tail(expand.(tail), acc, expand)
-        {{:|, meta, [head, tail]}, acc}
-      elem, acc ->
-        ensure_evaled_element(elem, acc)
-    end, acc, list)
+    :lists.mapfoldl(
+      fn
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please move the function out to a variable as well.

File: lib/elixir/lib/kernel.ex [hunk 9]
```diff
@@ -3563,13 +3678,15 @@ defmodule Kernel do
     end
   end
 
-  defp module_nesting([x | t1], [x | t2], acc, full),
-    do: module_nesting(t1, t2, [x | acc], full)
+  defp module_nesting([x | t1], [x | t2], acc, full), do: module_nesting(t1, t2, [x | acc], full)
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please use do-end blocks on all clauses for `module_nesting`.

File: lib/elixir/lib/kernel.ex [hunk 10]
```diff
@@ -3904,35 +4027,48 @@ defmodule Kernel do
...
         false ->
           quote do
             _ = @enforce_keys
+
             def __struct__(kv) do
-              :lists.foldl(fn {key, val}, acc ->
-                Map.replace!(acc, key, val)
-              end, @struct, kv)
+              :lists.foldl(
+                fn {key, val}, acc ->
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: I believe you could try to have all of this on a single line.  If it doesn't git, then extract the functino to a variable.

File: lib/elixir/lib/kernel.ex [hunk 11]
```diff
@@ -4486,29 +4635,33 @@ defmodule Kernel do
...
 
       # TODO: Raise on 2.0
       %{file: file, line: line} = __ENV__
+
       if is_list(funs) do
-        :elixir_errors.warn(line, file,
+        :elixir_errors.warn(
+          line,
+          file,
           "passing a list to Kernel.defdelegate/2 is deprecated, " <>
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please extract the error message to a variable.

File: lib/elixir/lib/kernel.ex [hunk 12]
```diff
@@ -1672,29 +1682,40 @@ defmodule Kernel do
...
+
             %_{__exception__: true} = other ->
-              :erlang.raise :error, other, stacktrace
+              :erlang.raise(:error, other, stacktrace)
+
             other ->
-              message = "reraise/2 expects a module name, string or exception as the first argument, got: #{inspect other}"
-              :erlang.error ArgumentError.exception(message)
+              message =
+                "reraise/2 expects a module name, string or exception as the first argument, got: #{
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Please try this:

```elixir
message = <<
  "reraise/2 expects a module name, string or exception",
  "as the first argument, got: #{inspect other}"
>>
```

**Comment by josevalim** [comment]
@fabrik42 review done! That's definitely a big one! I have added some comments, please let me know if you have any questions and don't forget to run the formatter after tackling them!

File: lib/elixir/lib/kernel.ex [hunk 13]
```diff
@@ -975,8 +976,9 @@ defmodule Kernel do
       #=> %{b: 1}
 
   """
-  @spec tl(nonempty_maybe_improper_list(elem, tail)) ::
-        maybe_improper_list(elem, tail) | tail when elem: term, tail: term
+  @spec tl(nonempty_maybe_improper_list(elem, tail)) :: maybe_improper_list(elem, tail) | tail
+        when elem: term,
+             tail: term
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self: figure out a way to put all entries on the same line after merged.

File: lib/elixir/lib/kernel.ex (hunk 1 above)
Comment (fabrik42) [reply lib/elixir/lib/kernel.ex]: I tried this, but even with do/end block it is still a couple chars do long. I kept the block for better readability though.

File: lib/elixir/lib/kernel.ex (hunk 2 above)
Comment (fabrik42) [reply lib/elixir/lib/kernel.ex]: Please see the comment above :)

**Comment by fabrik42** [comment]
Thanks for the review! I updated the pull request.

File: lib/elixir/lib/kernel.ex [hunk 14]
```diff
@@ -2378,37 +2415,41 @@ defmodule Kernel do
...
-           :erlang.element(1, expr) != :__MODULE__ do
+       when is_tuple(expr) and :erlang.element(1, expr) != :__aliases__ and
+              :erlang.element(1, expr) != :__MODULE__ do
     unnest(expr, [{:map, key} | acc], all_map?, kind)
   end
 
   defp unnest(other, [], _all_map?, kind) do
     raise ArgumentError,
-      "expected expression given to #{kind} to access at least one element, got: #{Macro.to_string other}"
+          "expected expression given to #{kind} to access at least one element, got: #{
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 15]
```diff
@@ -2623,22 +2683,25 @@ defmodule Kernel do
...
-               do: doc
+          with {_, doc} when doc_attr? <- Module.get_attribute(env.module, name, stack), do: doc
+
         try do
           :elixir_quote.escape(value, false)
         rescue
           ex in [ArgumentError] ->
-            raise ArgumentError, "cannot inject attribute @#{name} into function/macro because " <> Exception.message(ex)
+            raise ArgumentError,
+                  "cannot inject attribute @#{name} into function/macro because " <>
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 16]
```diff
@@ -2623,22 +2683,25 @@ defmodule Kernel do
...
           {val, _} -> val
         end
 
       false ->
         {escaped, _} = :elixir_quote.escape(stack, false)
+
         quote do
-          with {_, doc} when unquote(doc_attr?) <-
+          with {_, doc}
+               when unquote(doc_attr?) <-
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 17]
```diff
@@ -2746,17 +2810,21 @@ defmodule Kernel do
...
+        end
       end
-    end)
+    )
   end
 
   defp build_if(_condition, _arguments) do
-    raise(ArgumentError, "invalid or duplicate keys for if, only \"do\" " <>
-      "and an optional \"else\" are permitted")
+    message =
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 18]
```diff
@@ -3174,36 +3249,43 @@ defmodule Kernel do
...
             in_var(in_module?, left, &in_list(&1, head, tail, expand, list, in_module?))
+
           {[head | tail], {_, vars_values}} ->
             {vars, values} = :lists.unzip(:lists.reverse(vars_values))
+            is_in_list = &in_list(&1, head, tail, expand, list, in_module?)
+
             quote do
               {unquote_splicing(vars)} = {unquote_splicing(values)}
-              unquote(in_var(in_module?, left, &in_list(&1, head, tail, expand, list, in_module?)))
+
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 19]
```diff
@@ -3563,13 +3672,20 @@ defmodule Kernel do
...
-    do: {String.to_atom(<<"Elixir.", h::binary>>),
-          :elixir_aliases.concat(:lists.reverse([h | acc]))}
-  defp module_nesting(_, _, _acc, full),
-    do: {nil, full}
+  defp module_nesting([x | t1], [x | t2], acc, full) do
+    module_nesting(t1, t2, [x | acc], full)
+  end
+
+  defp module_nesting([], [h | _], acc, _full) do
+    {
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

File: lib/elixir/lib/kernel.ex [hunk 20]
```diff
@@ -4423,26 +4555,36 @@ defmodule Kernel do
...
+      Enum.map(expand_aliases(module, __CALLER__), fn
+        expanded when is_atom(expanded) ->
+          quote do
+            require unquote(expanded)
+            unquote(expanded).__using__(unquote(opts))
+          end
+
+        _otherwise ->
+          raise ArgumentError,
+                "invalid arguments for use, expected a compile time atom or alias, got: #{
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Note to self.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**